src/celeritas/optical/detail/PreGenAction.cc

```diff
@@ -0,0 +1,170 @@
+/*!
+ * Execute the action with host data.
+ */
+void PreGenAction::execute(CoreParams const& params, CoreStateHost& state) const
+{
+    this->execute_impl(params, state);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Execute the action with device data.
+ */
+void PreGenAction::execute(CoreParams const& params,
+                           CoreStateDevice& state) const
+{
+    this->execute_impl(params, state);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Generate optical distribution data post-step.
+ *
+ * The distributions are stored in separate Cerenkov and scintillation buffers
+ * indexed by the current buffer size plus the track slot ID. The data is
+ * compacted at the end of each step by removing all invalid distributions. The
+ * order of the distributions in the buffers is guaranteed to be reproducible.
+ */
+template<MemSpace M>
+void PreGenAction::execute_impl(CoreParams const& core_params,
+                                CoreState<M>& core_state) const
+{
+    size_type state_size = core_state.size();
+    StreamId stream = core_state.stream_id();
+    auto& buffer_size = storage_->size[stream.get()];
+    auto const& state = storage_->obj.state<M>(stream, state_size);
+
+    CELER_VALIDATE(buffer_size.cerenkov + state_size <= state.cerenkov.size(),
+                   << "insufficient capacity (" << state.cerenkov.size()
+                   << ") for buffered Cerenkov distribution data (total "
+                      "capacity requirement of "
+                   << buffer_size.cerenkov + state_size << ")");
+    CELER_VALIDATE(
+        buffer_size.scintillation + state_size <= state.scintillation.size(),
+        << "insufficient capacity (" << state.scintillation.size()
+        << ") for buffered scintillation distribution data (total "
+           "capacity requirement of "
+        << buffer_size.scintillation + state_size << ")");
+
+    // Generate the optical distribution data
+    this->pre_generate(core_params, core_state);
+
+    // Compact the buffers
+    buffer_size.cerenkov = this->remove_if_invalid(
+        state.cerenkov, buffer_size.cerenkov, state_size, stream);
+    buffer_size.scintillation = this->remove_if_invalid(
+        state.scintillation, buffer_size.scintillation, state_size, stream);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Launch a (host) kernel to generate optical distribution data post-step.
+ */
+void PreGenAction::pre_generate(CoreParams const& core_params,
+                                CoreStateHost& core_state) const
+{
+    TrackExecutor execute{
+        core_params.ptr<MemSpace::native>(),
+        core_state.ptr(),
+        detail::PreGenExecutor{properties_->host_ref(),
+                               cerenkov_->host_ref(),
+                               scintillation_->host_ref(),
```

Comment:
Actually this definitely won't work as is if only one of the processes is enabled... do you think it would be cleaner to just split this into two actions as @whokion suggested or try to fix it to work as a single action (@sethrj)?